### ISSUE TITLE
Embedding benchmark improvement via autonomous research agents

### ISSUE THREAD

**Issue by pierretokns** [pierretokns/agenthub#1]
## Goal

Use the autoresearch-mlx autonomous experimentation loop + agenthub coordination to train competitive embedding models on Apple Silicon, targeting specific MTEB subtasks where small models can win.

## Hardware

- M2 Ultra, 64GB unified memory
- MLX framework (native Apple Silicon)

### pplx-embed (Perplexity, Feb 2025)
- **Paper**: [arXiv 2602.11151](https://arxiv.org/abs/2602.11151)
- **Models**: pplx-embed-v1 (0.6B, 4B) + pplx-embed-context-v1 (0.6B, 4B)
- **HF collection**: [perplexity-ai/pplx-embed](https://huggingface.co/collections/perplexity-ai/pplx-embed)
- **Key technique**: Diffusion-continued pretraining on Qwen3 → multi-stage contrastive learning
- Diffusion pretraining enables bidirectional attention, making mean pooling work well
- 1024-dim embeddings, 32K context, INT8/BINARY quantization
- Record performance on ConTEB (contextual embeddings)
- Eliminates instruction tuning requirement

### "How to Train the Best Embedding Model" (Jack Morris / jxmo)
- **Blog post**: https://blog.jxmo.io/p/how-to-train-the-best-embedding-model
- **Tweet**: https://x.com/jxmnop/status/2031051636068782402
- **Core insight**: Embedding models plateau at ~10M training pairs because naive scaling introduces false negative label noise (collisions)
- **Solution**: Use LLMs to generate perfect labels via verification asymmetry — LLMs are much smarter than embedding models but more expensive
- **Training recipe**:
  1. Gather all available text pair datasets
  2. Filter top-k candidates per query using BM25 + existing embeddings (tractability)
  3. Label each (query, document) pair with an LLM to identify true negatives
  4. Train using exact softmax via coordinate ascent (not batch-level approximation)
- **CDE model**: ~10% improvement from embedding-based false negative filtering
- **Key constraint**: D² labeling problem — LLM labeling makes this a cost problem rather than a fundamental limitation

### Tier 1: High confidence
| Benchmark | Why | Approach |
|-----------|-----|----------|
| **MTEB STS** | Small models already competitive; low batch size friendly | Contrastive fine-tuning on sentence pairs, iterate on loss/pooling |
| **MTEB Pair Classification** | 0.6B models score 81+ | Well-understood fine-tuning recipe |
| **MTEB Clustering** | Instruction tuning shows "drastic improvement" | Autonomous iteration on training strategies |

### Tier 2: High value, more challenging
| Benchmark | Why | Approach |
|-----------|-----|----------|
| **MTEB Retrieval (domain-specific)** | 10-30% improvement with domain fine-tuning | Synthetic query-doc pairs via Claude, iterative hard negative mining |
| **ConTEB** | Perplexity set records; underexplored | Late chunking + document context |

### Tier 3: Differentiated moonshots
| Benchmark | Why | Approach |
|-----------|-----|----------|
| **Numerical Reasoning** | ALL models score ~54% (near chance) | Specialized numerical tokenization + targeted training |
| **MTEB Instruction Retrieval** | Average score 5.61 across all models | Floor so low that any focused work shows huge improvement |

### Adapt autoresearch-mlx loop for embeddings

```
Current: modify train.py → run 5 min → measure val_bpb → keep/discard
New:     modify train.py → run N min → measure MTEB subtask → keep/discard
```

### Base model options
- **Qwen3-0.6B** (mirrors pplx-embed approach)
- **ModernBERT-base** (149M, proven embedding backbone)
- **Snowflake Arctic-Embed-M** (110M, strong retrieval baseline)

### Training approach
1. Diffusion-style pretraining for bidirectional attention (pplx-embed insight)
2. Multi-stage contrastive learning (InfoNCE + hard negatives)
3. LLM-labeled training data to avoid false negative poisoning (jxmo insight)
4. Coordinate ascent / exact softmax instead of batch approximation
5. Iterative hard negative mining: train → mine harder negatives → retrain

### Key hyperparameters to iterate on
- Loss functions: InfoNCE, triplet, diffusion denoising
- Pooling: mean, CLS, weighted, attention
- Training data mixing and source stratification
- Hard negative mining threshold and strategy
- Architecture: attention patterns, embedding dimensions, MRL support
- Learning rate schedules, batch sizes (gradient accumulation to 2048+)

## Multi-Agent Coordination (via agenthub)

- **Agent 1**: STS / Pair Classification specialist
- **Agent 2**: Retrieval + hard negative mining specialist
- **Agent 3**: Clustering + instruction tuning specialist
- Agents share results via agenthub message board and git DAG

## Timeline

| When | What |
|------|------|
| Day 1 | Build embedding training loop in MLX, adapt autoresearch protocol, establish baselines |
| Night 1 | Launch autonomous agent loop |
| Days 2-4 | Monitor, adjust strategy, spin up multi-agent |
| Days 5-6 | Final push on best benchmarks |

## MTEB Landscape (for reference)

Current top models:
- Qwen3-Embedding-8B: 70.58 (overall MTEB)
- NV-Embed (8B): 69.32
- BGE-M3 (568M): ~63
- Nomic-embed-text-v1.5 (137M): ~59
- all-MiniLM-L6-v2 (23M): ~56

Small model gap is 7-14 points below 8B models — specific subtasks are much more tractable.
